modules/tinymce/src/core/main/ts/delete/TableDeleteAction.ts

```diff
@@ -28,120 +44,175 @@ interface TableCellRng {
-
-  return rng.collapsed ? Optional.none() : Optionals.lift2(startCell, endCell, tableCellRng).fold(
-    () => startCell.fold(
-      () => endCell.bind((endCell) =>
-        getClosestTable(endCell, isRoot).bind((table) =>
-          Arr.head(getTableCells(table)).map((startCell) => tableCellRng(startCell, endCell))
-        )
-      ),
-      (startCell) => getClosestTable(startCell, isRoot).bind((table) =>
-        Arr.last(getTableCells(table)).map((endCell) => tableCellRng(startCell, endCell))
-      )
-    ),
-    (cellRng: TableCellRng) => isWithinSameTable(isRoot, cellRng) ? Optional.none() : getCellRangeFromStartTable(cellRng, isRoot)
+const getCellRangeFromEndTable = (endCell: SugarElement<HTMLTableCellElement>, isRoot: IsRootFn): Optional<TableCellRng> =>
+  TableCellSelection.getClosestTable(endCell, isRoot).bind((table) =>
+    Arr.head(TableDeleteUtils.getTableCells(table)).map((startCell) => tableCellRng(startCell, endCell))
   );
-};
-
-const isWithinSameTable = (isRoot: IsRootFn, cellRng: TableCellRng) =>
-  getTableFromCellRng(cellRng, isRoot).isSome();
 
 const getTableSelectionFromCellRng = (cellRng: TableCellRng, isRoot: IsRootFn) =>
-  getTableFromCellRng(cellRng, isRoot).map((table) => tableSelection(cellRng, table, getTableCells(table)));
-
-const getTableSelection = (optCellRng: Optional<TableCellRng>, rng: Range, isRoot: IsRootFn) => optCellRng
-  .filter((cellRng) => isExpandedCellRng(cellRng) && isWithinSameTable(isRoot, cellRng))
-  .orThunk(() => partialSelection(isRoot, rng))
-  .bind((cRng) => getTableSelectionFromCellRng(cRng, isRoot));
+  getTableFromCellRng(cellRng, isRoot).map((table) => tableSelection(cellRng, table, TableDeleteUtils.getTableCells(table)));
+
+const getTableSelections = (cellRng: Optional<TableCellRng>, selectionDetails: SelectionDetails, rng: Range, isRoot: IsRootFn): Optional<TableSelections> => {
+  if (rng.collapsed || !cellRng.forall(isExpandedCellRng)) {
+    return Optional.none();
+  } else {
+    if (selectionDetails.isSameTable) {
+      const sameTableSelection = cellRng.bind((cRng) => getTableSelectionFromCellRng(cRng, isRoot));
+      return Optional.some({
+        start: sameTableSelection,
+        end: sameTableSelection
+      });
+    } else {
+      // Covers partial table selection (either start or end will have a tableSelection) and multitable selection (both start and end will have a tableSelection)
+      const startCell = getClosestCell(rng.startContainer, isRoot);
+      const endCell = getClosestCell(rng.endContainer, isRoot);
+      const startTableSelection = startCell.bind((start) => getCellRangeFromStartTable(start, isRoot)).bind((cRng) => getTableSelectionFromCellRng(cRng, isRoot));
+      const endTableSelection = endCell.bind((end) => getCellRangeFromEndTable(end, isRoot)).bind((cRng) => getTableSelectionFromCellRng(cRng, isRoot));
+      return Optional.some({
+        start: startTableSelection,
+        end: endTableSelection
+      });
+    }
+  }
```

Comment:
```suggestion
  } else if (selectionDetails.isSameTable) {
    const sameTableSelection = cellRng.bind((cRng) => getTableSelectionFromCellRng(cRng, isRoot));
    return Optional.some({
      start: sameTableSelection,
      end: sameTableSelection
    });
  } else {
    // Covers partial table selection (either start or end will have a tableSelection) and multitable selection (both start and end will have a tableSelection)
    const startCell = getClosestCell(rng.startContainer, isRoot);
    const endCell = getClosestCell(rng.endContainer, isRoot);
    const startTableSelection = startCell.bind((start) => getCellRangeFromStartTable(start, isRoot)).bind((cRng) => getTableSelectionFromCellRng(cRng, isRoot));
    const endTableSelection = endCell.bind((end) => getCellRangeFromEndTable(end, isRoot)).bind((cRng) => getTableSelectionFromCellRng(cRng, isRoot));
    return Optional.some({
      start: startTableSelection,
      end: endTableSelection
    });
  }
```
This doesn't need to be in a nested if/else block as it's all part of the single decision.

Reply:
Thanks for that. Should be good now.